File: banks2resources.py
#!/usr/bin/env python3

import os
import sys
from Unpacker import Unpacker

def read_byte(m):
    return ord(m.read(1))

def read_word(m):
    v = read_byte(m)
    v = v << 8 | read_byte(m)
    return v

def read_quad(m):
    v = read_word(m)
    v = v << 16 | read_word(m)
    return v

def read_mem_entry(memlist, n):
    memlist.seek(20*n)
    entry = {}
    entry["size"] = read_byte(memlist)
    entry["type"] = read_byte(memlist)
    entry["unknown_0x02"] = read_word(memlist) # unknown
    entry["unknown_0x04"] = read_word(memlist) # unknown
    entry["rankNum"] = read_byte(memlist)
    entry["bankId"] = read_byte(memlist)
    entry["bankOffset"] = read_quad(memlist)
    entry["unknown_0x0C"] = read_word(memlist) # unknown
    entry["packedSize"] = read_word(memlist)
    entry["unknown_0x10"] = read_word(memlist) # unknown
    entry["size"] = read_word(memlist)
    return entry


def read_msdos_memlist(memlist_filename):
    if not os.path.exists(memlist_filename):
        print (f"Memlist file was not found at: {memlist_filename}")
        sys.exit(-1)

    memlist = open(memlist_filename, "rb")
    mem_entries = []

    i = 0
    while True:
        entry = read_mem_entry(memlist, i)
        if entry["bankOffset"] == 0xFFFFFFFF:
            return mem_entries
        i += 1
        if entry["packedSize"] != 0:
            mem_entries.append(entry)


def get_type(i):
    if i==5:
        return "BYTECODE"
    else:
        return i

def main():
    if len(sys.argv) != 3:
        print (f"usage: {sys.argv[0]} <memlist.bin> <output_dir>\n")
        sys.exit(-1)

    memlist_filename = sys.argv[1]
    output_dir = sys.argv[2]
    if not os.path.exists(output_dir):
        os.mkdir(output_dir)

    directory = os.path.dirname(memlist_filename)
    entries = read_msdos_memlist(memlist_filename)
    for resource_index, entry in enumerate(entries):
        if entry != None:
            print (f"id:{entry['bankId']}\ttype:{get_type(entry['type'])}"
                   f"\toffset:{hex(entry['bankOffset'])}"
                   f"\tsize:{entry['packedSize']}/{entry['size']}")

        bank_file = os.path.join(directory, "bank%02x" % entry["bankId"])
        bank = open(bank_file, "rb")
        bank.seek(entry["bankOffset"])
        data = bank.read(entry["packedSize"])
        if entry["packedSize"] != entry["size"]:
            unpacker = Unpacker(data)
            success = unpacker.unpack(entry["packedSize"])

        if not success:
            print ("ERROR: Failed to unpack data.")
        else:
            bin_filename = os.path.join(output_dir,
                                        f"resource-0x{resource_index+1:02x}.bin")
            unpacker.buffer.seek(0)
            open(bin_filename, "wb").write(unpacker.buffer.read())
        bank.close()


if __name__ == "__main__":
    main()

File: awvm-tools/src/bin/banks2resources.rs
//! Rust port of `releases/common_data/banks2resources.py`.
//!
//! Reads `memlist.bin` from `<input_dir>` and extracts every resource
//! into `<output_dir>/resources/resource-0xNN.bin` — same filename
//! pattern, same per-entry stdout log line, as the Python reference.
//!
//! Usage:
//!     banks2resources <input_dir> <output_dir> [--uppercase]

use std::fs;
use std::path::PathBuf;
use std::process::ExitCode;

use awvm::bank;
use awvm::memlist::{self, MemEntry};

fn main() -> ExitCode {
    let mut args = std::env::args().skip(1);
    let input_dir: PathBuf = match args.next() {
        Some(s) => s.into(),
        None => {
            eprintln!("usage: banks2resources <input_dir> <output_dir> [--uppercase]");
            return ExitCode::from(2);
        }
    };
    let output_dir: PathBuf = match args.next() {
        Some(s) => s.into(),
        None => {
            eprintln!("usage: banks2resources <input_dir> <output_dir> [--uppercase]");
            return ExitCode::from(2);
        }
    };
    let mut uppercase = false;
    for arg in args {
        match arg.as_str() {
            "--uppercase" => uppercase = true,
            other => {
                eprintln!("unknown argument: {other}");
                return ExitCode::from(2);
            }
        }
    }

    let resources_dir = output_dir.join("resources");
    if let Err(e) = fs::create_dir_all(&resources_dir) {
        eprintln!("failed to create {}: {e}", resources_dir.display());
        return ExitCode::from(1);
    }

    let memlist_path = input_dir.join("memlist.bin");
    let memlist_bytes = match fs::read(&memlist_path) {
        Ok(b) => b,
        Err(e) => {
            eprintln!("failed to read {}: {e}", memlist_path.display());
            return ExitCode::from(1);
        }
    };

    let entries = match memlist::parse(&memlist_bytes) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("failed to parse memlist: {e}");
            return ExitCode::from(1);
        }
    };

    let mut had_failure = false;
    for (resource_index, entry) in entries.iter().enumerate() {
        log_entry(resource_index, entry);

        let bank_path = bank::bank_path(&input_dir, entry.bank_id, uppercase);
        if !bank_path.exists() {
            println!("Not found: {}", bank_path.display());
            continue;
        }

        match bank::read_resource(&input_dir, entry, uppercase) {
            Ok(data) => {
                let out = resources_dir.join(format!("resource-0x{:02x}.bin", resource_index));
                if let Err(e) = fs::write(&out, &data) {
                    eprintln!("failed to write {}: {e}", out.display());
                    had_failure = true;
                }
            }
            Err(bank::ReadError::SizeMismatch { expected, got }) => {
                // Python reference behaviour: "SHOULD BE %d ---- GOT %d" then sys.exit(-1).
                println!("SHOULD BE {expected} ---- GOT {got}");
                return ExitCode::from(255);
            }
            Err(e) => {
                eprintln!("resource 0x{resource_index:02x}: {e}");
                had_failure = true;
            }
        }
    }

    if had_failure { ExitCode::from(1) } else { ExitCode::SUCCESS }
}

fn log_entry(resource_index: usize, entry: &MemEntry) {
    // Match the Python reference's stdout exactly:
    //   resource:0x<idx>\tbankId:<id>\ttype:<TYPE>\toffset:0x<off>\tsize:0x<packed> / 0x<size>\tnext:0x<off+packed>
    let next = entry.bank_offset.wrapping_add(entry.packed_size as u32);
    println!(
        "resource:{resource}\tbankId:{bank}\ttype:{ty}\toffset:{off}\tsize:{packed} / {size}\tnext:{next}",
        resource = format!("0x{:x}", resource_index),
        bank = entry.bank_id,
        ty = memlist::type_name(entry.typ),
        off = format!("0x{:x}", entry.bank_offset),
        packed = format!("0x{:x}", entry.packed_size),
        size = format!("0x{:x}", entry.size),
        next = format!("0x{:x}", next),
    );
}
